indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java

```diff
@@ -749,131 +651,69 @@ private void handleStatus(final TaskStatus status)
-        }
-
-        // Add newly Added tasks to the queue
-        for (Task task : addedTasks) {
-          addTaskInternal(task);
-        }
-
-        log.info(
-            "Synced %d tasks from storage (%d tasks added, %d tasks removed).",
-            tasksSynced,
-            addedTasks.size(),
-            removedTasks.size()
-        );
-        requestManagement();
-      } else {
-        log.info("Not active. Skipping storage sync.");
-      }
-    }
-    catch (Exception e) {
-      log.warn(e, "Failed to sync tasks from storage!");
-      throw new RuntimeException(e);
-    }
-    finally {
-      giant.unlock();
-    }
-  }
-
-  private static Map<String, Task> toTaskIDMap(List<Task> taskList)
-  {
-    Map<String, Task> rv = new HashMap<>();
-    for (Task task : taskList) {
-      rv.put(task.getId(), task);
-    }
-    return rv;
-  }
-
-  private Map<String, Long> getDeltaValues(Map<String, Long> total, Map<String, Long> prev)
-  {
-    return total.entrySet()
-                .stream()
-                .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue() - prev.getOrDefault(e.getKey(), 0L)));
-  }
+    final Map<String, Task> newTasks = taskStorage.getActiveTasks().stream().collect(
+        Collectors.toMap(Task::getId, Function.identity())
+    );
+    final Map<String, Task> oldTasks = new HashMap<>(tasks);
+
+    // Calculate differences on IDs instead of Task Objects.
+    Set<String> commonIds = Sets.intersection(newTasks.keySet(), oldTasks.keySet());
+    for (String taskId : commonIds) {
+      newTasks.remove(taskId);
+      oldTasks.remove(taskId);
+    }
+    Collection<Task> addedTasks = newTasks.values();
+    Collection<Task> removedTasks = oldTasks.values();
+
+    // Add new tasks and clean up removed tasks
+    addedTasks.forEach(this::addTaskInternal);
+    removedTasks.forEach(this::removeTaskInternal);
+    log.info(
+        "Synced [%d] tasks from storage. Added [%d] tasks, removed [%d] tasks.",
+        newTasks.size(), addedTasks.size(), removedTasks.size()
+    );
 
-  public Map<String, Long> getSuccessfulTaskCount()
-  {
-    Map<String, Long> total = CollectionUtils.mapValues(totalSuccessfulTaskCount, AtomicLong::get);
-    synchronized (totalSuccessfulTaskCount) {
-      Map<String, Long> delta = getDeltaValues(total, prevTotalSuccessfulTaskCount);
-      prevTotalSuccessfulTaskCount = total;
-      return delta;
-    }
+    requestManagement();
   }
 
-  public Map<String, Long> getFailedTaskCount()
+  public Map<String, Long> getAndResetSuccessfulTaskCounts()
   {
-    Map<String, Long> total = CollectionUtils.mapValues(totalFailedTaskCount, AtomicLong::get);
-    synchronized (totalFailedTaskCount) {
-      Map<String, Long> delta = getDeltaValues(total, prevTotalFailedTaskCount);
-      prevTotalFailedTaskCount = total;
-      return delta;
-    }
+    Map<String, Long> total = new HashMap<>(datasourceToSuccessfulTaskCount);
+    datasourceToSuccessfulTaskCount.clear();
```

Comment:
This and following fns have a small race if the CHM was changed between the call to clear. There could be some alternatives, like iterating entries and using replaceAll, or remove() or merge() in a loop, that would allow a more atomic unload of the CHM? Or, maybe make the value in the map an AtomicLong, and CAS modify it to zero during unload?

Reply:
Yeah, I didn't like this either. Updated to iterate over keys and atomically remove them one by one.